pkg/provision/gce.go

```diff
@@ -0,0 +1,210 @@
+package provision
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"google.golang.org/api/compute/v1"
+	"google.golang.org/api/option"
+)
+
+// GCEProvisioner holds reference to the compute service to provision compute resources
+type GCEProvisioner struct {
+	gceProvisioner *compute.Service
+}
+
+// NewGCEProvisioner returns a new GCEProvisioner
+func NewGCEProvisioner(accessKey string) (*GCEProvisioner, error) {
+	gceService, err := compute.NewService(context.Background(), option.WithCredentialsJSON([]byte(accessKey)))
+	return &GCEProvisioner{
+		gceProvisioner: gceService,
+	}, err
+}
+
+// Provision provisions a new GCE instance as an exit node
+func (gce *GCEProvisioner) Provision(host BasicHost) (*ProvisionedHost, error) {
+	// instance auto restart on failure
+	autoRestart := true
+	instance := &compute.Instance{
+		Name:         host.Name,
+		MachineType:  fmt.Sprintf("zones/%s/machineTypes/%s", host.Additional["zone"], host.Plan),
+		CanIpForward: true,
+		Zone:         fmt.Sprintf("projects/%s/zones/us-central1-a", host.Additional["projectid"]),
+		Disks: []*compute.AttachedDisk{
+			{
+				AutoDelete: true,
+				Boot:       true,
+				DeviceName: host.Name,
+				Mode:       "READ_WRITE",
+				Type:       "PERSISTENT",
+				InitializeParams: &compute.AttachedDiskInitializeParams{
+					Description: "Inlets-Operator exit node",
+					DiskName:    host.Name,
+					DiskSizeGb:  10,
+					SourceImage: host.OS,
+				},
+			},
+		},
+		Metadata: &compute.Metadata{
+			Items: []*compute.MetadataItems{
+				{
+					Key:   "startup-script",
+					Value: &host.UserData,
+				},
+			},
+		},
+		Tags: &compute.Tags{
+			Items: []string{"http-server", "https-server", "inlets"},
+		},
+		Scheduling: &compute.Scheduling{
+			AutomaticRestart:  &autoRestart,
+			OnHostMaintenance: "MIGRATE",
+			Preemptible:       false,
+		},
+		NetworkInterfaces: []*compute.NetworkInterface{
+			{
+				AccessConfigs: []*compute.AccessConfig{
+					{
+						Type: "ONE_TO_ONE_NAT",
+						Name: "External NAT",
+					},
+				},
+				Network: "global/networks/default",
+			},
+		},
+		ServiceAccounts: []*compute.ServiceAccount{
+			{
+				Email: "default",
+				Scopes: []string{
+					compute.ComputeScope,
+				},
+			},
+		},
+	}
+
+	op, err := gce.gceProvisioner.Instances.Insert(host.Additional["projectid"], host.Additional["zone"], instance).Do()
+	if err != nil {
+		return nil, fmt.Errorf("could not provision GCE instance: %v", err)
+	}
+
+	status := ""
+
+	if op.Status == "RUNNING" {
+		status = ActiveStatus
+	}
+	return &ProvisionedHost{
+		ID:     constructCustomGCEID(host.Name, host.Additional["zone"], host.Additional["projectid"]),
+		Status: status,
+	}, nil
+
+}
+
+// CheckInletsFirewallRuleExists checks if the inlets firewall rule exists or not
+func (gce *GCEProvisioner) CheckInletsFirewallRuleExists(id string, inletsPort string) (bool, error) {
```

Comment:
Does this need to be public?

Reply:
No it doesn't need to be public. I initially thought of putting the logic of creation of resources, like the instance and firewall in the controller so it is more transparent for what we are "controlling" as the actual platform dependent code for the firewall creation is already inside the provision package. But I can do that too to keep it clean. :)